### PR TITLE
Add bilibili-official with version 1.1.2

### PR DESCRIPTION
**Pull request by wjlmelody**
**Important:** *Do not tick a checkbox if you haven’t performed its action.* Honesty is indispensable for a smooth review process.

_In the following questions `<cask>` is the token of the cask you're submitting._

Additionally, **if adding a new cask**:

- [x] Named the cask according to the [token reference](https://docs.brew.sh/Cask-Cookbook#token-reference).
- [x] Checked the cask was not [already refused](https://github.com/Homebrew/homebrew-cask/search?q=is%3Aclosed&type=Issues).
- [x] Checked the cask is submitted to [the correct repo](https://docs.brew.sh/Acceptable-Casks#finding-a-home-for-your-cask).
- [x] `brew audit --new-cask <cask>` worked successfully.
- [x] `brew install --cask <cask>` worked successfully.
- [x] `brew uninstall --cask <cask>` worked successfully.

### CODE REVIEW

File: Casks/bilibili-official.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  desc "Official bilibili video streaming and sharing platform"
+  homepage "https://app.bilibili.com/"
+
+  livecheck do
+    skip "No version information available"
+  end
+
+  auto_updates true
+
+  app "哔哩哔哩.app"
```
Comment (i0ntempest) [inline Casks/bilibili-official.rb:18]: I would remove this because auto update is disabled by default

**Comment by i0ntempest** [comment]
@bevanjkay There's an unofficial and no longer maintained Bilibili client that has taken the token `bilibili`. Should we just replace it with this official client?

**Comment by wjlmelody** [comment]
> @bevanjkay There's an unofficial and no longer maintained Bilibili client that has taken the token `bilibili`. Should we just replace it with this official client?

I support this replacement.

**Comment by SMillerDev** [comment]
Let's go with a second cask first (we can add a caveat to the old one informing users of the official cask). It's a better user experience than software being unexpectedly replaced on a `brew upgrade`

File: Casks/bilibili-official.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+  url "https://dl.hdslb.com/mobile/fixed/pc_electron_mac/bili_mac.dmg",
+      verified: "dl.hdslb.com/"
+  name "Bilibili-official"
+  name "哔哩哔哩官方客户端"
+  desc "Official bilibili video streaming and sharing platform"
+  homepage "https://app.bilibili.com/"
+
+  livecheck do
+    skip "No version information available"
```
Comment (zuisong) [inline Casks/bilibili-official.rb:13]:
```
  livecheck do
    url "http://api.bilibili.com/x/elec-frontend/update/latest-mac.yml"
    regex(/version: ([\d.]+)/i)
  end
```

livecheck  could be this

**Comment by ApolloZhu** [comment]
> Let's go with a second cask first (we can add a caveat to the old one informing users of the official cask). It's a better user experience than software being unexpectedly replaced on a `brew upgrade`

I agree that this will be good experience for the existing users who have installed `bilibili`. However, let me argue why we should at least:
1. start renaming the current `bilibili` cask to avoid confusing the more significant population of users who'd like to install the official client, and 
2. set the exact timeline/conditions for the official one to reclaim the token.

In my [previous communications](https://github.com/orgs/Homebrew/discussions/3287) I've forgotten how something I consider to be common knowledge may not be the case for others, so here's a little more detail on @bilibili (yes, this is their GitHub organization) to hopefully convince you it's a popular software/service that is what people would think of when they type `bilibili`. bilibili is a well known video sharing and live streaming platform among Chinese people from around the world. It is a publicly traded company on Nasdaq as [$BILI](https://www.nasdaq.com/symbol/bili), has 62.6K followers on [Twitter@bilibili_en](https://twitter.com/bilibili_en), and has at least **110 million** monthly active users per [investor presentation](https://ir.bilibili.com/static-files/dd6059b1-7376-4703-bdcd-6307f76249be) from 2019, and are considered by many to be China's YouTube (such as in [this article](https://www.economist.com/business/2021/03/25/bilibili-chinas-youtube-wants-to-be-its-netflix) from The Economist). I can't guarantee how many of them are using a Mac and would install the official client other than the 5 reacted in agreement to https://github.com/Homebrew/homebrew-cask/pull/124073#issuecomment-1132354362, but I think you can see how people are likely going to be confused to see a 3rd party client installed when they type in the trademarked name.

In comparison, the 3rd-party unofficial client only has about 1 download per day according to cask stats (28 in last 30 days, 65 in 90 days, and 310 in 365 days), its homepage https://typcn.com/legacy/bilimac/ can't be opened and says it's legacy software, there were no updates since September 2018 despite the significant number of open issues reporting bugs, and most of the functionalities don't work:

|Home Page Can't Contact Server|Download Manager Domain on Sale|
|--|--|
|||

Imagine the following scenario: I made a popular, 3rd party client for YouTube and named my project YouTube. Even though there are no official YouTube macOS client, it doesn't seem right for me to take the `youtube` token on Homebrew Cask and should instead be prefixed with a vendor name per [Potentially Misleading Name](https://docs.brew.sh/Cask-Cookbook#potentially-misleading-name). Similarly here,

> the token (`bilibili`) for a piece of unofficial software (https://github.com/typcn/bilibili-mac-client) that interacts with a popular service (i.e. bilibili/哔哩哔哩) would make it look official and the vendor is not authorised to use the name (**_i.e. not the trademark owner_**), [a prefix must be added](https://docs.brew.sh/Acceptable-Casks#forks-and-apps-with-conflicting-names) for disambiguation.

Therefore, in conclusion, renaming the existing cask is the right thing to do per the existing Homebrew rules and guidelines, will cause minimal interruptions for existing users due to the limited number of them and the condition of the software installed, and users who didn't read the cask descriptions and just follow their intuition can install the intended app without surprise.

**Comment by SMillerDev** [comment]
While that opinion might work in an ideal scenario, you're running into some cask limitation there:

Renaming casks can't happen without breaking the install for everyone who has the cask installed already. It'll simply stop updating without any warning.

The cleanest way forward is finishing this pull request and then adding a message to the existing `bilibili` cask to users to switch to this official one.

Then, when the downloads for the other cask drop to zero we can remove it.

**Comment by i0ntempest** [comment]
This could be another zoom/zoomus scenario. I’m not sure how the download counter works, does it count the people who install `bilibili`, see the message then immediately replace it with this official client? If it does I doubt it’ll drop to 0 any time soon.

**Comment by SMillerDev** [comment]
It won't be quick, but it'll make sure the people who need it are aware of the official build.

It's not the best option but literally any option is better than swapping out software people willingly installed for completely different software on an update. Such an act would gravely damage Homebrew's trustworthiness as a package manager. Imagine the outcry if we swapped out the Firefox cask for a Chrome build. This isn't much different.

**Comment by i0ntempest** [comment]
Okay I'll accept that. Now can we merge this?

**Comment by ApolloZhu** [comment]
I agree that we should merge this if it's ready. But let me ask a few questions about the replacement process later so we can be prepared.

> swapping out software people willingly installed for completely different software on an update ... would gravely damage Homebrew's trustworthiness

It's unfortunate that [Renaming a Formula](https://docs.brew.sh/Rename-A-Formula) only works with homebrew-core and I totally agree with the above point.

> Then, when the downloads for the other cask drop to zero we can remove it.

There have been many pull requests that rename casks due to misleading names, such as #105712, #105689, #105688, but none of them have clear information about download stats. For a one that does -- #71506:

Please see pull request 71506 for alt text

... it broke the install for more than 7000 people (not 310). I'm okay with that we are moving to a stricter standard that only 0 downloads means okay to replace, but I think we **need clarification on if it's 0 downloads in 30 days, 90 days, or 365 days and document that in the official documentation**. Note that this standard  **guarantees that we break the install for more people** that installed `bilibili-official` when the replacement happens, so that's just something to keep in mind.

Considering that removing a cask entirely is a rather drastic approach and it prevents users who do want to use it, how about in addition to adding `bilibili-official`, **we duplicate the definition of `bilibili` to one with proper vendor prefix (e.g. `typcn-bilibili`), and add a note to `bilibili` (in addition to the one informing about `bilibili-official`) so that existing users have more time to migrate?**

**Review by SMillerDev — DISMISSED** [review]

**Comment by SMillerDev** [comment]
> we duplicate the definition of bilibili to one with proper vendor prefix (e.g. typcn-bilibili), and add a note to bilibili (in addition to the one informing about bilibili-official) so that existing users have more time to migrate?

Yeah, sounds good

File: Casks/bilibili-official.rb [hunk 3]
```diff
@@ -4,23 +4,23 @@
...
   auto_updates true
 
   app "哔哩哔哩.app"
 
   zap trash: [
-    "~/Library/Application Support/bilibili/",
-    "~/Library/Saved Application State/com.bilibili.bilibiliPC.savedState/",
+    "~/Library/Application Support/bilibili",
     "~/Library/Preferences/com.bilibili.bilibiliPC.plist",
+    "~/Library/Saved Application State/com.bilibili.bilibiliPC.savedState",
```
Comment (i0ntempest) [inline Casks/bilibili-official.rb:24]: One more zap item:
`~/Library/Logs/bilibili/`

**Review by miccal — APPROVED** [review]

**Pull request merged**